src/main/java/picard/illumina/ExtractIlluminaBarcodes.java

```diff
@@ -223,23 +235,78 @@ protected int doWork() {
         }
 
         LOG.info("Processing with " + numProcessors + " PerTileBarcodeExtractor(s).");
-        final ExecutorService pool = Executors.newFixedThreadPool(numProcessors);
-
-        // TODO: This is terribly inefficient; we're opening a huge number of files via the extractor constructor and we never close them.
-        final List<PerTileBarcodeExtractor> extractors = new ArrayList<PerTileBarcodeExtractor>(factory.getAvailableTiles().size());
-        for (final int tile : factory.getAvailableTiles()) {
-            final PerTileBarcodeExtractor extractor = new PerTileBarcodeExtractor(
-                    tile,
-                    getBarcodeFile(tile),
-                    barcodeToMetrics,
-                    noMatchMetric,
-                    factory,
-                    MINIMUM_BASE_QUALITY,
-                    MAX_NO_CALLS,
-                    MAX_MISMATCHES,
-                    MIN_MISMATCH_DELTA
-            );
-            extractors.add(extractor);
+        final ThreadPoolExecutor pool = new ThreadPoolExecutorWithExceptions(numProcessors);
+
+        final List<PerTileBarcodeExtractor> extractors = new ArrayList<>(factory.getAvailableTiles().size());
+
+        if (USE_NEW_CONVERTER) {
+            File laneDir = new File(BASECALLS_DIR, IlluminaFileUtil.longLaneStr(LANE));
+
+            File[] cycleDirs = IOUtil.getFilesMatchingRegexp(laneDir, IlluminaFileUtil.CYCLE_SUBDIRECTORY_PATTERN);
+
+            //CBCLs
+            List<File> cbcls = new ArrayList<>();
+            Arrays.asList(cycleDirs)
```

Comment:
Some refactoring needed here to avoid code duplication. E.g. the code on line 261 could be moved into a static method in `AbstractIlluminaPositionFileReader`.